[#28] 동네 인증 Controller 구현 및 Service 수정

Srltas opened the pull request:
## 📌 Summary
- 동네 인증 결과를 반환하는 `NeighborhoodVerificationResponse`를 추가하고, 새로운 동네 인증 Controller를 구현했습니다.

## 📝 Description
1. Controller 구현
   - POST로 동네 인증 요청 구현
   - Session에 user id 값을 얻어서 사용하는 방식으로 구현
2. Service, Repository 리팩토링
   - NeighborhoodVerificationResponse를 반환하도록 수정
   - `@Service`, `@Repository`를 붙여 Spring의 의존주입을 받을 수 있도록 수정
3. 동네 인증 테스트 코드 작성
   - Parameterize를 사용해 여러 개의 파라미터 값을 검증할 수 있도록 구현
 
## 📚 References
- [동네 인증 API 문서](https://github.com/f-lab-edu/run-together/wiki/run%E2%80%90together-API)

## ✅ Checklist
- [x] 새로운 기능이나 수정된 기능에 대해 충분한 테스트를 작성했습니다.
- [x] 코딩 스타일 가이드를 준수했습니다.
- [x] 문서(주석, README 등)가 필요하다면 업데이트했습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/NeighborhoodVerificationController.java:
```diff
@@ -0,0 +1,47 @@
+	@PostMapping("/{neighborhoodName}/verification")
+	public ResponseEntity<NeighborhoodVerificationResponse> verifyNeighborhood(
+		@PathVariable String neighborhoodName,
+		@RequestBody @Validated NeighborhoodVerificationRequest neighborhoodVerificationRequest) {
```
Validated 와 Valid 의 차이가 뭔지 학습 하고 이 부분 어떤걸 사용하는게 적절할지 고민 해 보세요

Srltas replied:
- **`@Valid`**: 객체의 필드나 메서드 파라미터에 대한 유효성 검사를 수행할 때 사용하는 자바 표준 어노테이션입니다. Spring에서는 이를 지원하기 위해 기본적으로 Hibernate Validator 같은 Bean Validation 구현체를 사용합니다.
- **`@Validated`**: Spring에서 제공하는 확장된 유효성 검사 어노테이션입니다. 그룹 검증(Group Validation) 기능이 추가되어 특정 검증 그룹을 나눠서 검사할 수 있다는 특징이 있습니다.

`@Valid`는 자바 표준 Bean Validation API에 의존성하기 때문에 Spring과의 결합도가 낮습니다. 반면, `@Validation`은 Spring의 유효성 검사 확장 기능을 사용하므로, Spring에 의존성이 있습니다.

현재 코드에서 수행하는 검증은 단순히 DTO의 필드들이 올바르게 입력되었는지 확인하는 정도이고, `@Valid`를 사용하면 Spring에 대한 의존성을 줄여 결합도를 낮출 수 있기 때문에, `@Valid`를 사용하는 것이 적합하다고 생각합니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/adapter/in/NeighborhoodVerificationController.java:
```diff
@@ -0,0 +1,47 @@
+import lombok.RequiredArgsConstructor;
+
+@RequiredArgsConstructor
+@Validated
```
Validated 가 왜 필요한지 고민 한 번 해 보세요.

Srltas replied:
해당 위치에 Validated를 설정하면 클래스의 모든 메서드가 유효성 검증을 수행할 수 있도록 설정하는 것입니다.
이미 메서드의 파라미터의 Validated를 붙여 사용하고 있기 때문에 의미가 없는 코드라고 생각합니다.
수정하도록 하겠습니다.

f-lab-bradley commented on src/main/java/com/srltas/runtogether/application/mappper/LocationMapper.java:
```diff
@@ -0,0 +1,18 @@
+import com.srltas.runtogether.application.port.in.NeighborhoodVerificationCommand;
+import com.srltas.runtogether.domain.model.location.Location;
+
+public class LocationMapper {
```
이건 그냥 꿀 팁인데, 만약에 Dto 간 변환시 각 필드에 대한 변수명이 같으면 지피티한테 대충 두개 코드 던져주고 매퍼 생성해줘 하면 알아서 생성 해 줘요 ㅋㅋㅋ

f-lab-bradley reviewed (commented):
main 에 리베이스 한 번 해주세요

Srltas committed:
build: springboot validation 추가

Srltas committed:
feature: 동네인증 Controller 구현

- Session에서 user id만 있는 상태로 가정해 Controller 구현
- Controller 응답을 위해 application에 response DTO 생성
- Controller 응답을 위해 application의 Service 수정
- Spring의 의존주입을 받기 위해 @Service, @Repository 추가

Srltas committed:
test: 동네인증 Controller 테스트 코드

- parameterized를 이용해 성공 케이스만 추가

Srltas commented on src/main/java/com/srltas/runtogether/adapter/in/NeighborhoodVerificationController.java:
```diff
@@ -0,0 +1,36 @@
+	@PostMapping("/neighborhood/verification")
+	public ResponseEntity<NeighborhoodVerificationResponse> verifyNeighborhood(
+		@RequestBody @Valid NeighborhoodVerificationRequest neighborhoodVerificationRequest,
+		@SessionAttribute(name = "login_user_id", required = false) Long userId) {
```
User ID를 Session에서 확인하는 방식으로 수정했습니다. 
하지만, Session에 User ID 한 개만 넣고 사용하는 방식으로 만들었는데 이 부분에 대해서 실제로는 어떻게 사용하는, 이 방식의 문제가 있다면 어떤 것이 있는지 피드백을 받고 싶습니다!..

f-lab-bradley replied:
저라면 아래 처럼 할 것 같아요.

1. 사용자 토큰을 통해 Filter 에서 Session Storage 로 부터 매핑된 사용자를 확인하여 사용자 식별
2. HttpSession 에 사용자 정보 저장
3. Controller 에서 HttpSession 으로 부터 사용자 정보 로드하여 사용

위의 흐름대로 구현 해 보고 대충 어떻게 돌아가는건지 파악이 되면 추후 Spring Security 를 도입 하는 것도 고려 할 수 있습니다~

--

그리고 이건 인증에 관련 된 것 이고 Controller 구현과는 조금 거리가 있으므로 해당 부분은 새로 PR 을 만들면 좋을 것 같네요

Srltas commented:
> main 에 리베이스 한 번 해주세요

리베이스 완료 했습니다!

Srltas committed:
refactor: DTO 필드 유효성 검사 추가

- 위도, 경도 최소, 최대값 설정
- neighborhood 정수값만 사용할 수 있도록 설정

f-lab-bradley reviewed (approved):
사용자 인증 부분 바로 이어서 진행 해 주세요
